### DISCUSSION TITLE
@Inject within interceptor and feature

### DISCUSSION THREAD

**Discussion by dinosath** [quarkiverse/quarkus-cxf#312] [Q&A] [answered]
I would like to know if it is possible to use @Inject annotation within interceptors and features

**Comment by shumonsharif** [comment, marked as answer]
Hi @dinosath I assume you are referring to custom interceptors and features; the [docs](https://quarkiverse.github.io/quarkiverse-docs/quarkus-cxf/dev/interceptors-and-features.html) clearly state that:

"Both feature and interceptor classes will be attempted to be loaded via CDI first, and if no CDI beans are available, then the constructor with no parameters will be invoked to instantiate each class."

I would suggest you review the following link thoroughly:
https://quarkus.io/guides/cdi-reference

I did not test it, but in theory, adding both `@Startup` and `@Singleton` annotations to your custom interceptor or feature classes should do the trick for you.  Please let us know in case that doesn't work.

**Reply by dinosath** [reply]
Thanks for the reply. It helped a lot. I was missing `@Startup` annotation and thus it didn't work.

**Comment by arolfes** [comment]
Hi @dinosath Hi @shumonsharif ,

I was stumbling about the same issue. In my case when I set the `@Singleton` I always see this log message:

```
================================================================================
CDI: programmatic lookup problem detected
-----------------------------------------
At least one bean matched the required type and qualifiers but was marked as unused and removed during build

Stack frame: io.quarkiverse.cxf.CxfClientProducer.addToCols(CxfClientProducer.java:145)
Required type: class info.novatec.examples.soapclient.quarkus.soap.SoapHeadersInterceptor
Required qualifiers: [@javax.enterprise.inject.Default()]
Removed beans:
	- CLASS bean  [types=[org.apache.cxf.interceptor.Interceptor<org.apache.cxf.binding.soap.SoapMessage>, interface org.apache.cxf.binding.soap.interceptor.SoapInterceptor, org.apache.cxf.phase.PhaseInterceptor<org.apache.cxf.binding.soap.SoapMessage>, org.apache.cxf.phase.AbstractPhaseInterceptor<org.apache.cxf.binding.soap.SoapMessage>, class org.apache.cxf.binding.soap.interceptor.AbstractSoapInterceptor, class info.novatec.examples.soapclient.quarkus.soap.SoapHeadersInterceptor], qualifiers=null]
```

With the additional annotation `@Unremovable` the exception was gone but the properties where not injected...

I had to add `quarkus.arc.remove-unused-beans=framework` to `application.properties` and then it if properly injected.

**Reply by arolfes** [reply]
Is it possible to add such example to the documentation?

**Reply by shumonsharif** [reply]
Hi @arolfes I apologize, I'm not sure how I missed your comments here from a couple months ago.  What you're describing may possibly be a bug in the extension.  Would you be able to share a reproducer?  Assuming there's not an issue in the extension, then this topic would be a welcome addition to the docs - please feel free to submit a PR.

**Reply by arolfes** [reply]
Hi @shumonsharif ,

No Problem. With my knowledge now I believe it was my fault.

Let me try to reproduce it and if I can then i will share the example and explain it.

**Reply by arolfes** [reply]
One question
Should I use latest quarkus and latest graalvm with java 17?

**Reply by shumonsharif** [reply]
Hi @arolfes yes, latest quarkus and quarkus-cxf would be good.  As far as java 17, I don't believe that is fully supported yet - so Java 11 would possibly be better.  Thanks, and please keep us posted!
